NuRadioReco/detector/RNO_G/rnog_detector.py

```diff
@@ -0,0 +1,1572 @@
+
+from NuRadioReco.detector.RNO_G.db_mongo_read import Database, _convert_astro_time_to_datetime
+import NuRadioReco.framework.base_trace
+from NuRadioReco.utilities import units
+
+from radiotools import helper
+from scipy import interpolate
+from functools import wraps
+import astropy.time
+
+import datetime
+import numpy as np
+import collections
+import pickle
+import json
+import re
+import copy
+import bson
+import lzma
+
+
+def _json_serial(obj):
+    """JSON serializer for objects not serializable by default json code"""
+
+    if isinstance(obj, (datetime.datetime, datetime.date)):
+        return obj.isoformat()
+    elif isinstance(obj, bson.objectid.ObjectId):
+        return str(obj)
+    elif isinstance(obj, Response):
+        pass
+    else:
+        raise TypeError ("Type %s not serializable" % type(obj))
+
+
+def _keys_not_in_dict(d, keys):
+    """ Checks sequentially if a list of `keys` is in a dictionary.
+
+    Example:
+    keys = ["key1", "key2"]
+
+    Returns False if d["key1"]["key2"] exsits, True otherwise.
+    """
+    d_tmp = d
+    for key in keys:
+        try:
+            if key not in d_tmp:
+                return True
+        except KeyError:
+            return True
+
+        d_tmp = d_tmp[key]
+
+    return False
+
+
+def _check_detector_time(method):
+    @wraps(method)
+    def _impl(self, *method_args, **method_kwargs):
+        self.get_detector_time()  # this will raise an error if time is not set
+        return method(self, *method_args, **method_kwargs)
+    return _impl
+
+
+class Detector():
+    def __init__(self, database_connection='RNOG_test_public', log_level=logging.INFO, over_write_handset_values={},
+                 database_time=None, always_query_entire_description=True, detector_file=None,
+                 select_stations=None):
+        """
+
+        Parameters
+        ----------
+
+        database_connection : str (Default: 'RNOG_test_public')
+            Allows to specify database connection. Passed to mongo-db interface.
+
+        log_level : `logging.LOG_LEVEL` (Default: logging.INFO)
+            Defines verbosity level of logger. Other options are: `logging.WARNING`, `logging.DEBUG`, ...
+
+        over_write_handset_values : dict (Default: {})
+            Overwrite the default values for the manually set parameter which are not (yet) implemented in the database.
+            (Default: {}, the acutally default values for the parameters in question are defined below)
+
+        database_time : `datetime.datetime`
+            Set database time which is used to select the primary measurement. By default (= None) the database time
+            is set to now (time the code is running) to select the measurement which is now primary.
+
+        always_query_entire_description : bool (Default: True)
+            If True, query the entire detector describtion all at once when calling Detector.update(...) (if necessary).
```

Comment:
Maybe this is on me, but what happens if this is `False`?

Reply:
If that is false only the station_list collection is queried with `update` all other informations such as position or signal chain responses are queried when requested (and then also added to the buffer). One example is this function here:
```
    @_check_detector_time
    def get_absolute_position(self, station_id):
        """
        Get the absolute position of a specific station (relative to site)

        Parameters
        ----------

        station_id: int
            the station id

        Returns
        -------

        pos: np.array(3,)
            3-dim array of absolute station position in easting, northing and depth wrt. to snow level at
            time of measurement
        """
        self.logger.debug(
            f"Requesting station position for station {station_id}.")

        if _keys_not_in_dict(self.__buffered_stations, [station_id, "station_position"]):

            if _keys_not_in_dict(self.__buffered_stations, [station_id, "id_position"]):
                raise KeyError(
                    f"\"id_position\" not in buffer for station {station_id}. Did you call det.update(..)")

            station_position_id = self.__buffered_stations[station_id]["id_position"]
            self.logger.debug(
                f"Query position for id \"{station_position_id}\"")

            station_position = self.__db.get_position(
                position_id=station_position_id)
            self.__buffered_stations[station_id]["station_position"] = station_position

        return np.array(self.__buffered_stations[station_id]["station_position"]["position"])
```